internal/registry/registry.go

```diff
@@ -0,0 +1,206 @@
+			lastOSVersion = configs[i].OSVersion
+			result = append(result, newImage(r.baseURL, images[i], configs[i]))
+		}
+	}
+	if len(result) == 0 {
+		return nil, fmt.Errorf("image tag %s not found for platform %s", tag, platform)
+	}
+	return result[len(result)-1], nil
+}
+
+func (r *registry) getImageManifests(ctx context.Context, tag string) ([]*imageManifestV1, error) {
+	header := http.Header{}
+	header.Add("Accept", acceptImageIndexV1)
+	header.Add("Accept", acceptImageManifestV1)
+
+	url := fmt.Sprintf("%s/manifests/%s", r.baseURL, tag)
+	body, mediaType, err := r.httpClient.Get(ctx, url, header)
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()         //nolint
+	b, err := io.ReadAll(body) // fully read the response
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case strings.Contains(acceptImageIndexV1, mediaType):
+		index := imageIndexV1{}
+		if err = json.Unmarshal(b, &index); err != nil {
+			return nil, fmt.Errorf("error unmarshalling image index from %s: %w", url, err)
+		}
+		return r.getMultiPlatformManifests(ctx, &index)
+	case strings.Contains(acceptImageManifestV1, mediaType):
+		manifest := imageManifestV1{}
+		if err = json.Unmarshal(b, &manifest); err != nil {
+			return nil, fmt.Errorf("error unmarshalling image manifest from %s: %w", url, err)
+		}
+		manifest.URL = url
+		return []*imageManifestV1{&manifest}, nil
+	default:
+		return nil, fmt.Errorf("unknown mediaType %s from %s: %w", mediaType, url, err)
+	}
+}
+
+func (r *registry) getMultiPlatformManifests(ctx context.Context, index *imageIndexV1) ([]*imageManifestV1, error) {
+	var manifests = make([]*imageManifestV1, len(index.Manifests))
+	for i, ref := range index.Manifests {
+		url := fmt.Sprintf("%s/manifests/%s", r.baseURL, ref.Digest)
+		manifest := imageManifestV1{}
+		if err := r.httpClient.GetJSON(ctx, url, ref.MediaType, &manifest); err != nil {
+			return nil, fmt.Errorf("error getting image ref for platform %v: %w", ref.Platform, err)
+		}
+		manifest.URL = url
+		manifests[i] = &manifest
+	}
+	return manifests, nil
+}
+
+func (r *registry) getImageConfigs(ctx context.Context, images []*imageManifestV1) ([]*imageConfigV1, error) {
+	var configs = make([]*imageConfigV1, len(images))
+	for i, image := range images {
+		if !strings.Contains(acceptImageConfigV1, image.Config.MediaType) {
+			return nil, fmt.Errorf("invalid config media type in image %v", image)
+		}
+		url := fmt.Sprintf("%s/blobs/%s", r.baseURL, image.Config.Digest)
+		config := imageConfigV1{}
+		err := r.httpClient.GetJSON(ctx, url, image.Config.MediaType, &config)
+		if err != nil {
+			return nil, fmt.Errorf("error getting image config from %s: %w", url, err)
+		}
+		configs[i] = &config
+	}
+	return configs, nil
+}
+
+func (r *registry) ReadFilesystemLayer(ctx context.Context, layer *internal.FilesystemLayer, readFile internal.ReadFile) error {
+	header := http.Header{}
+	header.Add("Accept", layer.MediaType)
+	body, _, err := r.httpClient.Get(ctx, layer.URL, header)
```

Comment:
I believe body.Close() should be called since httpClient.Get returns Reader*Closer*

Reply:
yep totally right. missed that!